### PR TITLE
feat: log all critical bounces

### PR DESCRIPTION
**Pull request by mantariksh**
## Problem

The current bounce collection and alarm system work as follows:
1. When a bounce notification comes in, update `hasBounced` in the Bounce collection for each email recipient.
2. If `hasBounced` is true for all the email recipients AND `hasAlarmed` is false, log a message containing `CRITICAL BOUNCE` and set `hasAlarmed` to true. If `hasAlarmed` is already true, do nothing.
3. In CloudWatch, filter for messages containing `CRITICAL BOUNCE`. Every 5min, if there is at least 1 message containing `CRITICAL BOUNCE` in the last 5min, activate the alarm.

In other words, we do not log every single critical bounce. This is a mechanism to limit the rate of alarms; if there are critical bounces for one form ID, but lots of people submitting, we will still get only ONE alarm every 3 hours. This is because every 3 hours, the document in the Bounce collection will expire, upon which `hasAlarmed` will be set to its default of false. Once one critical bounce comes in, `hasAlarmed` will be set to true. Subsequent critical bounces will see that `hasAlarmed` is true and not log anything, hence will not activate the alarm again.

The issue with this system is that we are unable to track the number of critical bounces over time, since we only send one log message per critical bounce per form every 3 hours. This is problematic for two reasons:
1. We cannot set an alarm based on the volume of critical bounces.
2. We cannot track the rate of critical bounces over time. This is further exacerbated by the fact that the logs are sent to a special CloudWatch log group which has a 2-month TTL. This means that we cannot track bounces which are older than 2 months.

## Solution

1. Log ALL critical bounces, including those for which alarms have already been activated. We can then set an alarm based on the total number of critical bounces over a period of time.
2. Log email notifications to the main log group, which allows us to search for a year's worth of bounce info for admins. The exception here is email notifications for Email Confirmations sent to form-fillers, since there are privacy concerns regarding keeping metadata on citizens for an extended period of time. Hence Email Confirmation notifications will still be logged to the short-term log group.

The following refactors and configuration changes were also made:
- `bounce.server.model` was moved inside `src/app/modules/bounce` for better grouping of related code.
- Tests for the `bounce` model were inlined in `src/app/modules/bounce/__tests__`. A couple of configuration changes were made to accommodate this:
    - `tsconfig.build.json` was updated so that `__tests__` directories are not compiled.
    - `jest.config.js` was updated so that files in the `__tests__` directories are not included when calculating test coverage.

### Pre-release
- [x] Staging critical bounces alarm has been updated to search for logs in the main log group.
- [x] Admin email notifications go to the main log group and not `formsg-email-notifications-staging`. 

### Post-release
- [ ] Production critical bounces alarm has been updated to search for logs in the main log group.
- [ ] Production dashboard widgets have been updated.

### CODE REVIEW

**Comment by shuli-ogp** [comment]
Think you left out something in the Solution write-up "Log all critical bounces, including those for which {?}"

File: src/app/modules/bounce/bounce.service.ts [hunk 1]
```diff
@@ -129,18 +137,34 @@ const logCriticalBounce = (
...
+      },
+    })
+  }
   const latestBounces = Bounce.fromSnsNotification(notification)
   if (!latestBounces) return
   const formId = latestBounces.formId
   const oldBounces = await Bounce.findOne({ formId })
   if (oldBounces) {
     oldBounces.merge(latestBounces, notification)
-    logCriticalBounce(oldBounces, formId, notification)
```
Comment (shuli-ogp) [inline src/app/modules/bounce/bounce.service.ts:140]: What is the reason for removing formId from this log?

Comment (mantariksh) [reply src/app/modules/bounce/bounce.service.ts:140]: We're not removing the form ID from the logs - the form ID is contained within `notification`, so this was just a refactor where we extract `formID` from `notification` in the `logCriticalBounce` function instead of passing it as a redundant parameter

File: src/app/modules/bounce/__tests__/bounce-test-helpers.ts [hunk 2]
```diff
@@ -26,6 +26,7 @@ const makeEmailNotification = (
   formId: ObjectId,
   submissionId: ObjectId,
   recipientList: string[],
+  emailType: 'Admin (response)' | 'Email confirmation',
```
Comment (karrui) [inline src/app/modules/bounce/__tests__/bounce-test-helpers.ts:29]: nit: Make this into an enum? Since it's used more than once :P

File: src/app/modules/bounce/bounce.service.ts [hunk 3]
```diff
@@ -101,20 +108,21 @@ export const isValidSnsRequest = async (
...
     logger.warn({
-      type: 'CRITICAL BOUNCE',
-      formId,
-      recipients: bounceDoc.bounces.map((emailInfo) => emailInfo.email),
-      // We know for sure that critical bounces can only happen because of bounce
-      // notifications, so this casting is okay
-      bounceInfo: (notification as IBounceNotification).bounce,
+      message: 'CRITICAL BOUNCE',
+      meta: {
+        action: 'Received critical bounce',
```
Comment (karrui) [inline src/app/modules/bounce/bounce.service.ts:117]: shouldn't action be `isValidSnsRequest`? We should follow a convention haha. Maybe the message can be "Received critical bounce" and add "CRITICAL BOUNCE" as `meta.bounceType` or something

File: src/app/modules/bounce/bounce.service.ts [hunk 4]
```diff
@@ -101,20 +108,21 @@ export const isValidSnsRequest = async (
...
 const logCriticalBounce = (
   bounceDoc: IBounceSchema,
-  formId: string,
   notification: IEmailNotification,
 ): void => {
-  if (
-    !bounceDoc.hasAlarmed &&
-    bounceDoc.bounces.every((emailInfo) => emailInfo.hasBounced)
-  ) {
+  if (bounceDoc.bounces.every((emailInfo) => emailInfo.hasBounced)) {
```
Comment (karrui) [inline src/app/modules/bounce/bounce.service.ts:113]: Maybe we can make a type guard using this conditional, so you don't need to do `notification as IBounceNotification` casting below

```ts
const isBounceNotification = (doc): doc is IBounceNotification => bounceDoc.bounces.every((emailInfo) => emailInfo.hasBounced)
```

File: src/app/modules/bounce/bounce.service.ts [hunk 5]
```diff
@@ -129,18 +137,34 @@ const logCriticalBounce = (
...
+  if (
+    extractHeader(notification, EMAIL_HEADERS.emailType) ===
+    EMAIL_TYPES.emailConfirmation
+  ) {
+    shortTermLogger.info(notification)
+  } else {
+    logger.info({
+      message: 'Email notification',
+      meta: {
+        action: 'Received notification from SNS',
```
Comment (karrui) [inline src/app/modules/bounce/bounce.service.ts:153]: Same here, action should be the method invocation name, in this case `updateBounces`.

**Review by karrui — APPROVED** [review]
lgtm, with exception of the logging parameters. `meta.action` should always be the calling function

File: src/app/modules/bounce/bounce.service.ts [hunk 6]
```diff
@@ -129,18 +137,34 @@ const logCriticalBounce = (
...
   const oldBounces = await Bounce.findOne({ formId })
   if (oldBounces) {
     oldBounces.merge(latestBounces, notification)
-    logCriticalBounce(oldBounces, formId, notification)
+    logCriticalBounce(oldBounces, notification)
     await oldBounces.save()
   } else {
-    logCriticalBounce(latestBounces, formId, notification)
+    logCriticalBounce(latestBounces, notification)
     await latestBounces.save()
```
Comment (arshadali172) [inline src/app/modules/bounce/bounce.service.ts:168]: I'm not sure what is the best way to do this, but I think that updating the `hasAlarmed` key should not be the responsibility of `logCriticalBounce`

```
if (oldBounces) {
    oldBounces.merge(latestBounces, notification)
    logCriticalBounce(oldBounces, notification)
    bounceDoc.hasAlarmed = true
    await oldBounces.save()
} else {
    logCriticalBounce(latestBounces, notification)
    bounceDoc.hasAlarmed = true
    await latestBounces.save()
}
```

File: src/app/modules/bounce/bounce.service.ts [hunk 7]
```diff
@@ -129,18 +137,34 @@ const logCriticalBounce = (
 export const updateBounces = async (body: ISnsNotification): Promise<void> => {
   const notification: IEmailNotification = JSON.parse(body.Message)
   // This is the crucial log statement which allows us to debug bounce-related
-  // issues, as it logs all the details about deliveries and bounces
-  logger.info(notification)
+  // issues, as it logs all the details about deliveries and bounces. Email
+  // confirmation info goes to the short-term log group so we do not store
+  // form fillers' information for too long, and everything else goes into the
+  // main log group.
+  if (
```
Comment (arshadali172) [inline src/app/modules/bounce/bounce.service.ts:144]: nit but wouldn't it be better if this `if ... else` was in its own helper? Similar to `logCriticalBounce()`

File: src/app/modules/bounce/__tests__/bounce-test-helpers.ts (hunk 2 above)
Comment (arshadali172) [reply src/app/modules/bounce/__tests__/bounce-test-helpers.ts:29]: can use the enum declared in `mail.ts` constants file

File: src/app/modules/bounce/__tests__/bounce-test-helpers.ts [hunk 8]
```diff
@@ -61,9 +62,16 @@ export const makeBounceNotification = (
   recipientList: string[] = [],
   bouncedList: string[] = [],
   bounceType: 'Transient' | 'Permanent' = 'Permanent',
```
Comment (arshadali172) [inline src/app/modules/bounce/__tests__/bounce-test-helpers.ts:64]: technically not part of this PR, but could we convert this to an enum too as part of the constants file or types file

File: src/app/modules/bounce/__tests__/bounce.service.spec.ts [hunk 9]
```diff
@@ -138,9 +145,11 @@ describe('updateBounces', () => {
       email,
       hasBounced: false,
     }))
-    expect(mockLogger.info).toHaveBeenCalledWith(
-      JSON.parse(notification.Message),
-    )
+    expect(mockLogger.info.mock.calls[0][0]).toMatchObject({
```
Comment (arshadali172) [inline src/app/modules/bounce/__tests__/bounce.service.spec.ts:148]: first time looking at these tests. Now I'm disappointed in myself for not mocking the logger as part of my tests :/

File: src/app/modules/bounce/bounce.util.ts [hunk 10]
```diff
@@ -13,7 +9,7 @@ export const extractHeader = (
   body: IEmailNotification,
   header: string,
 ): string => {
-  return get(body, 'mail.headers').find(
+  return body.mail.headers.find(
```
Comment (arshadali172) [inline src/app/modules/bounce/bounce.util.ts:12]: what's wrong with using lodash `get`? I think the fix here is that the default should be set for `get` to be `[]`?

File: src/app/modules/bounce/bounce.util.ts [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+/**
+ * Whether a bounce notification says a given email has bounced.
+ * @param bounceInfo Bounce notification from SNS
+ * @param email Email address to check
+ */
+export const hasEmailBounced = (
+  bounceInfo: IBounceNotification,
+  email: string,
+): boolean => {
+  return bounceInfo.bounce.bouncedRecipients.some(
```
Comment (arshadali172) [inline src/app/modules/bounce/bounce.util.ts:26]: similarly here. what is wrong with using `get`?

File: src/app/modules/bounce/bounce.model.ts [hunk 12]
```diff
@@ -58,26 +59,6 @@ const BounceSchema = new Schema<IBounceSchema>({
...
-// Whether a bounce notification says a given email has bounced
-const hasEmailBounced = (
-  bounceInfo: IBounceNotification,
-  email: string,
-): boolean => {
-  return get(bounceInfo, 'bounce.bouncedRecipients').some(
-    (emailInfo) => emailInfo.emailAddress === email,
-  )
-}
-
```
Comment (arshadali172) [inline src/app/modules/bounce/bounce.model.ts:80]: line 76 below:
`const bounces: ISingleBounce[] = get(snsInfo, 'mail.commonHeaders.to')`
Please add `[]` as a default for `get`?

File: src/app/modules/bounce/bounce.service.ts (hunk 6 above)
Comment (shuli-ogp) [reply src/app/modules/bounce/bounce.service.ts:168]: Perhaps in line 162, since we should be updating hasAlarmed regardless of whether there are oldBounces

```
  if (!latestBounces) return
  const formId = latestBounces.formId
  const oldBounces = await Bounce.findOne({ formId })
>>bounceDoc.hasAlarmed = true
  if (oldBounces) {
    oldBounces.merge(latestBounces, notification)
    logCriticalBounce(oldBounces, notification)
    await oldBounces.save()
  } else {
    logCriticalBounce(latestBounces, notification)
    await latestBounces.save()
  }

```

**Review by shuli-ogp — APPROVED** [review]

File: src/app/modules/bounce/bounce.service.ts (hunk 4 above)
Comment (liangyuanruo) [reply src/app/modules/bounce/bounce.service.ts:113]: alternatively, introduce an instance method `bounceDoc.allEmailsBounced()` to the Domain so that the business logic is clearly represented and elevated in importance

Comment (mantariksh) [reply src/app/modules/bounce/bounce.service.ts:113]: the problem here is that we are inferring the type of one object based on another object, i.e. we are inferring that the notification must be a bounce notification because the bounce document shows a critical bounce. I solved this by adding an explicit type check for the notification type, but will explore more options in the PR for auto deactivation and auto emailing, e.g. investigating canonical ways to implement DTOs

File: src/app/modules/bounce/bounce.service.ts (hunk 6 above)
Comment (mantariksh) [reply src/app/modules/bounce/bounce.service.ts:168]: absolutely. this key is going to be repurposed in a subsequent PR anyway, currently it's just dangling. I've replaced this with a TODO linking to the "auto email form admin" issue.

File: src/app/modules/bounce/__tests__/bounce.service.spec.ts (hunk 9 above)
Comment (mantariksh) [reply src/app/modules/bounce/__tests__/bounce.service.spec.ts:148]: lol no la it's just that the log statement is particularly important here, because our alarms depend on it

File: src/app/modules/bounce/__tests__/bounce-test-helpers.ts (hunk 8 above)
Comment (mantariksh) [reply src/app/modules/bounce/__tests__/bounce-test-helpers.ts:64]: have added as a TODO!

File: src/app/modules/bounce/bounce.util.ts (hunk 10 above)
Comment (mantariksh) [reply src/app/modules/bounce/bounce.util.ts:16]: Yeah so the issue is that once you use `_.get`, you lose the typing - everything gets converted to `any`. But you're right that there is still a danger because we are parsing a POST notification from an external source, which is why all the processing is wrapped in a try-catch in the controller (at least until we implement FP).

File: src/app/modules/bounce/bounce.util.ts (hunk 11 above)
Comment (mantariksh) [reply src/app/modules/bounce/bounce.util.ts:30]: as above

File: src/app/modules/bounce/bounce.util.ts (hunk 10 above)
Comment (mantariksh) [reply src/app/modules/bounce/bounce.util.ts:16]: In fact once TypeScript strict mode turns on, we can't use _.get anymore

File: src/app/modules/bounce/bounce.model.ts (hunk 12 above)
Comment (mantariksh) [reply src/app/modules/bounce/bounce.model.ts:80]: as above
